bin/jsii-release-maven

```diff
@@ -193,27 +226,110 @@ cat > ${release_pom} <<HERE
+    # override an existing version. Otherwise, fail!
+    if [ ${PIPESTATUS[0]} -ne 0 ]; then
+        if cat ${release_output} | grep "does not allow updating artifact" | grep -q ".pom"; then
+            echo "⚠️ Artifact already published. Skipping"
+        else
+            error "Release failed"
+        fi
+    fi
+}
+
+deploy_non_central() {
+    release_output="${workdir}/release-output.txt"
+    for pom in ${poms}; do
+        $mvn --settings=${mvn_settings} deploy -f ${pom} -DaltDeploymentRepository=${server_id}::default::${MAVEN_REPOSITORY_URL} | tee ${release_output}
+
+        # If release failed, check if this was caused because we are trying to publish
+        # the same version again, which is not an error. The magic string "409 Conflict"
+        # indicates that we are trying to
+        # override an existing version. Otherwise, fail!
+        if [ ${PIPESTATUS[0]} -ne 0 ]; then
+            if cat ${release_output} | grep -q "409 Conflict"; then
+                echo "⚠️ Artifact already published. Skipping"
+            else
+                error "Release failed"
+            fi
+        fi
+    done
+}
+
+main() {
+    cd "${1:-"dist/java"}"
+
+    server_id="${MAVEN_SERVER_ID:-"${CENTRAL_SERVER}"}"
+    if [[ "${server_id}" == "${CENTRAL_SERVER}" ]]; then
+        is_central=true
     else
-        echo "❌ Release failed"
-        exit 1
+        is_central=false
+    fi
+
+    if [[ -n "${MAVEN_GPG_PRIVATE_KEY_FILE:-}" ]] || [[ -n "${MAVEN_GPG_PRIVATE_KEY:-}" ]]; then
+        is_signed=true
+    else
+        is_signed=false
     fi
-fi
 
-echo "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
-echo "✅ All Done!"
+    validate_parameters
+
+    if [[ -n "${MAVEN_DRYRUN:-}" ]]; then
+        echo "==========================================="
+        echo "            🏜️ DRY-RUN MODE 🏜️"
+        echo "==========================================="
+        mvn="echo mvn"
+        dry_run=true
+    else
+        mvn=mvn
+        dry_run=false
+    fi
+
+    if [[ "${is_signed}" == true ]]; then
+        import_gpg_key
+    fi
+
+    poms="$(find . -name '*.pom')"
+    if [ -z "${poms}" ]; then
+        error "No JARS to publish: no .pom files found under $PWD"
+    fi
+
+    if [[ "${is_central}" == true ]]; then
+        echo "📦 Publishing to Maven Central"
+    else
+        echo "📦 Publishing to ${server_id}"
+    fi
+    
+    staging=$(mktemp -d)
+    workdir=$(mktemp -d)
+    echo ${workdir}
+
+    create_maven_settings
+
+    if [[ "${is_signed}" == true ]]; then
+        # This currently only works with Maven Central
+        sign_artifacts
+    fi
+
+    if [[ "${is_central}" == true ]]; then
+        deploy_central
+    else
+        deploy_non_central
+    fi
+
+    echo "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
+    echo "✅ All Done!"
+}
+
+main "$@"; exit
```

Comment:
Why is `exit` needed here?

Reply:
It's a safeguard against the script being modified while executing. (See https://unix.stackexchange.com/questions/331837/how-to-read-the-whole-shell-script-before-executing-it).

I'm making an effort to learn more about bash rather than just do "what works", but I'm fine removing some of the extra tweaks.

Reply:
TIL